File: ast/expression/increment.go
package expression

import (
	"github.com/panda-io/micro-panda/ast"
	"github.com/panda-io/micro-panda/ast/core"
	"github.com/panda-io/micro-panda/ast/types"
)

type Increment struct {
	ExpressionBase
	Expression ast.Expression
}

func (i *Increment) Validate(c ast.Context, expected core.Type) {
	i.Const = false
	i.Expression.Validate(c, expected)
	if i.Expression.IsConstant() {
		c.Error(i.GetPosition(), "expect variable")
	}
	if !types.IsInteger(i.Expression.Type()) {
		c.Error(i.GetPosition(), "expect integer expression")
	}
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/expression/Increment.java
package com.github.panda_io.micro_panda.ast.expression;

import com.github.panda_io.micro_panda.ast.type.Type;
import com.github.panda_io.micro_panda.ast.Context;

public class Increment extends Expression {
    public Expression expression;
    
	public boolean isLvalue() {
		return false;
	}
    
    public void validate(Context context, Type expected) {
        this.constant = false;
        this.expression.validate(context, expected);
        if (this.expression.type == null) {
            return;
        }
        if (this.expression.isConstant()) {
            context.addError(this.expression.getOffset(), "expect variable");
        }
        if (!this.expression.type.isInteger()) {
            context.addError(this.expression.getOffset(), "expect integer expression");
        }
    }
}
